# Midterm Review: Python Programming Essentials

This notebook provides a summary of key Python concepts covered in Sessions 1 through 5, to help you prepare for your midterm.

## 1. Python Fundamentals (from Session 1)

### 1.1. Variables

*   Named containers for storing data values.
*   Assignment: `variable_name = value`.
*   Naming Rules: Must start with a letter (a-z, A-Z) or an underscore (`_`). Can contain alphanumeric characters and underscores. Case-sensitive (`age`, `Age`, `AGE` are different).

In [ ]:
message = "Hello, Reviewer!"
count = 10
pi_value = 3.14159
is_ready = True
print(message, count, pi_value, is_ready)

Hello, Reviewer! 10 3.14159 True


### 1.2. Data Types

*   **Integer (`int`):** Whole numbers (e.g., `100`, `-5`).
*   **Float (`float`):** Numbers with a decimal point (e.g., `3.14`, `-0.001`).
*   **String (`str`):** Sequence of characters, for text (e.g., `"Hello"`, `'Python123'`).
*   **Boolean (`bool`):** Truth values, `True` or `False` (case-sensitive).
*   Python uses **dynamic typing**: you don't declare a variable's type.
*   The `type()` function shows a variable's data type.

In [ ]:
num_int = 100
num_float = 25.5
text_str = "Midterm"
flag_bool = False
print(type(num_int), type(num_float), type(text_str), type(flag_bool))

<class 'int'> <class 'float'> <class 'str'> <class 'bool'>


### 1.3. Basic Input/Output: `print()` and `input()`

*   **`print()`:** Displays output to the console.
    *   Can print multiple items: `print("Value:", x)`.
*   **`input(prompt_message)`:** Pauses execution, waits for user to type input, and returns the input as a **string**.
    *   Example: `name = input("Enter your name: ")`.

In [ ]:
print("This is a print example.")
# user_name = input("Enter your name: ") # Example, commented out for non-interactive run
user_name = "TestUser" # Placeholder for review notebook
print(f"Hello, {user_name}!") # Using f-string for formatted output

This is a print example.
Hello, TestUser!


### 1.4. Comments

*   Notes in your code ignored by Python.
*   Start with `#` for single-line comments.
*   Used to explain code or temporarily disable lines.

In [ ]:
# This is a full-line comment
x = 5 # This is an inline comment
print(x)
# print("This line won't execute")

5


### 1.5. Numerical Operations & Type Conversion

*   **Arithmetic Operators:** `+` (addition), `-` (subtraction), `*` (multiplication), `/` (float division), `//` (integer/floor division), `%` (modulus/remainder), `**` (exponentiation).
*   **Type Conversion:**
    *   `int(value)`: Converts to integer (e.g., `int("123")`, `int(3.9)` which truncates to `3`).
    *   `float(value)`: Converts to float (e.g., `float("3.14")`, `float(10)` which becomes `10.0`).
    *   `str(value)`: Converts to string (e.g., `str(100)`, `str(3.14)`).
    *   Crucial when using `input()` for numerical calculations.

In [ ]:
a = 10
b = 3
print(f"{a} + {b} = {a + b}")
print(f"{a} / {b} = {a / b}")
print(f"{a} // {b} = {a // b}")
print(f"{a} % {b} = {a % b}")
print(f"2 ** 3 = {2**3}")

num_str = "42" # Simulating input()
num_int = int(num_str) # Convert string to integer
print(f"String '{num_str}' converted to int: {num_int}, its square: {num_int ** 2}")

10 + 3 = 13
10 / 3 = 3.3333333333333335
10 // 3 = 3
10 % 3 = 1
2 ** 3 = 8
String '42' converted to int: 42, its square: 1764


## 2. Strings (from Session 2.1)

### 2.1. Creating Strings

*   Enclosed in single (`'`) or double (`"`) quotes.
*   **Escape Characters:** `\` allows special characters:
    *   `\'`: Single quote
    *   `\"`: Double quote
    *   `\\`: Backslash
    *   `\n`: Newline
    *   `\t`: Tab
*   **Multi-line Strings:** Use triple quotes (`'''` or `"""`) to span multiple lines and preserve formatting.

In [ ]:
my_string1 = "Hello, Python!"
my_string2 = 'It\'s a review with a newline:\nand a tab\t here.'
multi_line_string = """This is a
multi-line string."""
print(my_string1)
print(my_string2)
print(multi_line_string)

Hello, Python!
It's a review with a newline:
and a tab	 here.
This is a
multi-line string.


### 2.2. String Concatenation and f-Strings

*   **Concatenation (`+`):** Joins strings. Non-strings must be converted to `str()` first.
    *   `"Hello" + " " + "World"` results in `"Hello World"`.
*   **f-Strings (Formatted String Literals):** A concise way to embed expressions inside string literals. Prefix string with `f` or `F`.
    *   `name = "Alice"; age = 30; f"My name is {name} and I am {age}."`

In [ ]:
greeting = "Welcome"
name = "Student"
year = 2024

# Concatenation
full_greeting_concat = greeting + " " + name + "!"
print(full_greeting_concat)

# f-string
score = 95
full_greeting_fstring = f"{greeting}, {name}! Your score is {score}."
print(full_greeting_fstring)

Welcome Student!
Welcome, Student! Your score is 95.


### 2.3. Indexing and Slicing

*   **Indexing:** Access individual characters. Zero-based (`string[0]` is first char). Negative indexing (`string[-1]` is last char).
*   **Slicing:** Extract a subsequence (substring). `string[start:end:step]`.
    *   `start`: Included (default: beginning).
    *   `end`: Excluded (default: end).
    *   `step`: Interval (default: 1).

In [ ]:
text = "Python Rocks"
print(f"Original: {text}")
print(f"First char: {text[0]}")
print(f"Last char: {text[-1]}")
print(f"Slice [0:6]: {text[0:6]}") # or text[:6]
print(f"Slice [7:]: {text[7:]}")
print(f"Slice every 2nd char: {text[::2]}")
print(f"Reversed: {text[::-1]}")

Original: Python Rocks
First char: P
Last char: s
Slice [0:6]: Python
Slice [7:]: Rocks
Slice every 2nd char: Pto oc
Reversed: skcoR nohtyP


### 2.4. Common String Methods (from Session 2.1)

*   `len(string)`: Returns the length of the string.
*   `substring in string`: Boolean check if `substring` exists in `string`.
*   `string.find(substring)`: Returns the starting index of the first occurrence of `substring`. Returns -1 if not found.
*   `string.rfind(substring)`: Returns the starting index of the last occurrence of `substring`. Returns -1 if not found.
*   `string.replace(old, new)`: Returns a *new* string with all occurrences of `old` replaced by `new`. The original string is unchanged.

In [ ]:
s = "Midterm Review Session"
print(f"s: {s}")
print(f"Length: {len(s)}")
print(f"'Review' in s: {'Review' in s}")
print(f"Index of 'Review': {s.find('Review')}")
print(f"Index of last 'e': {s.rfind('e')}")
s_replaced = s.replace("Session", "Time")
print(f"Replaced 'Session' with 'Time': {s_replaced}")
print(f"Original s is unchanged: {s}")

s: Midterm Review Session
Length: 22
'Review' in s: True
Index of 'Review': 8
Index of last 'e': 18
Replaced 'Session' with 'Time': Midterm Review Time
Original s is unchanged: Midterm Review Session


## 3. Conditional Statements (from Session 2.2)

Conditional statements allow programs to make decisions and execute different code paths based on whether specific conditions are `True` or `False`.

### 3.1. `if`, `elif`, `else`

*   `if condition:`: Executes block if `condition` is `True`.
*   `elif another_condition:`: (Else if) Executes block if previous `if`/`elif` conditions were `False` and `another_condition` is `True`.
*   `else:`: Executes block if all preceding `if`/`elif` conditions were `False`.
*   Indentation is crucial for defining code blocks.

### 3.2. Comparison and Logical Operators

*   **Comparison Operators:** `==` (equal), `!=` (not equal), `>` (greater than), `<` (less than), `>=` (greater or equal), `<=` (less or equal).
*   **Logical Operators:** `and` (both true), `or` (at least one true), `not` (inverts boolean).

In [ ]:
temperature = 25
is_raining = False

if temperature > 30:
    print("It's very hot!")
elif temperature > 20 and not is_raining:
    print("It's nice and sunny.")
elif temperature < 10 or is_raining:
    print("It's cold or raining. Stay warm!")
else:
    print("Weather is moderate.")

It's nice and sunny.


### 3.3. Nested Conditions

You can place `if`/`elif`/`else` statements inside the blocks of other conditional statements for more complex logic.

In [ ]:
age = 20
has_ticket = True

if age >= 18:
    print("Adult detected.")
    if has_ticket:
        print("Welcome to the event!")
    else:
        print("You need a ticket to enter.")
else:
    print("You are too young for this event.")

Welcome to the event!


## 4. Data Structures

### 4.1. Lists (from Session 3.1)

*   Ordered, mutable (changeable) collection of items. Defined with `[]`.
*   Can contain items of various data types, including other lists.

#### 4.1.1. Indexing and Slicing (similar to strings)

In [ ]:
colors = ["red", "green", "blue", "yellow"]
print(f"Original: {colors}")
print(f"First color: {colors[0]}")
print(f"Last color: {colors[-1]}")
print(f"Slice [1:3]: {colors[1:3]}")

# Modify element
colors[1] = "lime"
print(f"Modified list (colors[1] = 'lime'): {colors}")

Original: ['red', 'green', 'blue', 'yellow']
First color: red
Last color: yellow
Slice [1:3]: ['green', 'blue']
Modified list (colors[1] = 'lime'): ['red', 'lime', 'blue', 'yellow']


#### 4.1.2. Common List Methods & Operations

*   `list.append(item)`: Adds `item` to the end.
*   `list.insert(index, item)`: Inserts `item` at `index`.
*   `list.pop(index)`: Removes and returns item at `index` (default: last item). 
*   `list.remove(value)`: Removes the first occurrence of `value`. Raises `ValueError` if not found.
*   `len(list)`: Returns the number of items.
*   `+` (concatenation): `list1 + list2` creates a new list.
*   `*` (repetition): `list * n` creates a new list with items repeated `n` times.
*   `item in list`: Boolean check for membership.

In [ ]:
numbers = [1, 2, 3]
print(f"Numbers: {numbers}")

numbers.append(4)
print(f"Appended 4: {numbers}")

numbers.insert(1, 1.5)
print(f"Inserted 1.5 at index 1: {numbers}")

popped_item = numbers.pop() 
print(f"Popped item (last): {popped_item}. List is now: {numbers}")

numbers.remove(1.5)
print(f"Removed 1.5: {numbers}")

print(f"Length: {len(numbers)}")
print(f"Concatenated with [5, 6]: {numbers + [5, 6]}")
print(f"Repeated twice: {numbers * 2}")
print(f"Is 2 in numbers? {2 in numbers}")

Numbers: [1, 2, 3]
Appended 4: [1, 2, 3, 4]
Inserted 1.5 at index 1: [1, 1.5, 2, 3, 4]
Popped item (last): 4. List is now: [1, 1.5, 2, 3]
Removed 1.5: [1, 2, 3]
Length: 3
Concatenated with [5, 6]: [1, 2, 3, 5, 6]
Repeated twice: [1, 2, 3, 1, 2, 3]
Is 2 in numbers? True


### 4.2. Dictionaries (from Session 3.2)

*   Collection of **key-value pairs**. Defined with `{}`: `my_dict = {key1: value1, key2: value2}`.
*   Keys must be unique and immutable (e.g., strings, numbers, tuples).
*   Values can be of any data type and can be duplicated.
*   Mutable.
*   Unordered (before Python 3.7) or insertion-ordered (Python 3.7+).

#### 4.2.1. Accessing, Adding, Modifying, Deleting Key-Value Pairs

*   **Access value:** `my_dict[key]` (Raises `KeyError` if key not found).
*   **Add new pair / Modify existing value:** `my_dict[key] = new_value`.
*   **Delete pair:** `del my_dict[key]` (Raises `KeyError` if key not found).

In [ ]:
student = {"name": "Alice", "age": 20, "major": "CS"}
print(f"Student: {student}")

print(f"Name: {student['name']}")

student["age"] = 21 # Modify existing value
student["year"] = 3  # Add new key-value pair
print(f"Updated student (age to 21, added year 3): {student}")

del student["major"]
print(f"After deleting 'major': {student}")

Student: {'name': 'Alice', 'age': 20, 'major': 'CS'}
Name: Alice
Updated student (age to 21, added year 3): {'name': 'Alice', 'age': 21, 'major': 'CS', 'year': 3}
After deleting 'major': {'name': 'Alice', 'age': 21, 'year': 3}


#### 4.2.2. Checking Key Existence & Other Operations

*   `key in my_dict`: Boolean check if `key` exists in the dictionary.
*   `my_dict.keys()`: Returns a view object displaying a list of all the keys.
*   `len(my_dict)`: Returns the number of key-value pairs.

In [ ]:
config = {"host": "localhost", "port": 8080}
print(f"Config: {config}")

print(f"Is 'port' a key? {'port' in config}")
print(f"Is 'user' a key? {'user' in config}")

print(f"Keys: {config.keys()}")
print(f"Length: {len(config)}")

Config: {'host': 'localhost', 'port': 8080}
Is 'port' a key? True
Is 'user' a key? False
Keys: dict_keys(['host', 'port'])
Length: 2


## 5. Loops

Loops allow code to be executed repeatedly.

### 5.1. `for` Loops (from Session 4.1)

*   Used for iterating over a sequence (like a list, tuple, dictionary, string) or other iterable objects.
*   Syntax: `for item in sequence:`

#### 5.1.1. Iterating over Sequences

In [ ]:
fruits = ["apple", "banana", "cherry"]
for fruit in fruits:
    print(fruit.capitalize()) # Example of doing something with each item

for char in "loop":
    print(char)

Apple
Banana
Cherry
l
o
o
p


#### 5.1.2. `range()` Function

Generates a sequence of numbers, often used with `for` loops.
*   `range(stop)`: Generates numbers from `0` up to (but not including) `stop`.
    *   Example: `range(5)` -> `0, 1, 2, 3, 4`.
*   `range(start, stop)`: Generates numbers from `start` up to `stop`.
    *   Example: `range(2, 5)` -> `2, 3, 4`.
*   `range(start, stop, step)`: Generates numbers from `start` up to `stop`, incrementing by `step`.
    *   Example: `range(1, 10, 2)` -> `1, 3, 5, 7, 9`.

In [ ]:
print("range(4):")
for i in range(4):
    print(i, end=" ")
print("\nrange(1, 5):")
for i in range(1, 5):
    print(i, end=" ")
print("\nrange(0, 10, 3):")
for i in range(0, 10, 3):
    print(i, end=" ")
print()

range(4):
0 1 2 3 
range(1, 5):
1 2 3 4 
range(0, 10, 3):
0 3 6 9 


### 5.2. `while` Loops (from Session 4.2)

*   Repeatedly executes a block of code as long as a given condition remains `True`.
*   Syntax: `while condition:`
*   **Important:** The loop body must contain code that can eventually make the `condition` `False` to avoid infinite loops.

In [ ]:
count = 0
while count < 3:
    print(f"While loop iteration: {count}")
    count += 1 # Crucial update step
print("While loop finished.")

While loop iteration: 0
While loop iteration: 1
While loop iteration: 2
While loop finished.


### 5.3. `for` vs `while`

*   **`for` loop:** Use when you know the number of iterations beforehand (e.g., iterating over all items in a list, or using `range()`).
*   **`while` loop:** Use when the number of iterations is unknown and depends on a condition becoming false (e.g., user input, waiting for an event).

### 5.4. `break` and `continue` (from Session 4.2)

*   **`break` Statement:** Immediately terminates the entire innermost loop (`for` or `while`). Execution continues at the first statement after the loop.
*   **`continue` Statement:** Immediately skips the rest of the code in the current iteration and jumps to the beginning of the next iteration of the loop.

In [ ]:
print("Break example (stops at 3):")
for i in range(10):
    if i == 3:
        print("Breaking loop now.")
        break
    print(i)

print("\nContinue example (skips 2):")
for i in range(5):
    if i == 2:
        print(f"Skipping iteration for {i}")
        continue
    print(i)

Break example (stops at 3):
0
1
2
Breaking loop now.

Continue example (skips 2):
0
1
Skipping iteration for 2
3
4


## 6. Functions (from Session 5.1)

A function is a named block of reusable code designed to perform a specific task.

### 6.1. Defining and Calling Functions

*   **Definition:** Uses the `def` keyword.
  ```python
  def function_name(parameter1, parameter2, ...):
      """Optional docstring explaining the function."""
      # Code block
      # ...
      return result # Optional
  ```
*   **Calling:** Execute the function by using its name followed by parentheses `()` containing arguments.
  `function_name(argument1, argument2, ...)`

### 6.2. Parameters and Arguments

*   **Parameters:** Variables listed inside the parentheses in the function definition. They are placeholders for the values that will be passed to the function.
*   **Arguments:** The actual values that are passed to the function when it is called.

### 6.3. `return` Statement

*   Used to send a value (or result) back from the function to the caller.
*   A function can have multiple `return` statements (e.g., in different `if`/`elif` branches).
*   If a function doesn't have an explicit `return` statement that returns a value, or if it has `return` without an expression, it implicitly returns `None`.

#### `return` vs. `print` within a Function

*   **`print()`:** Displays information to the console/output. It's for showing things to the user. It does *not* send a value back to the part of the code that called the function for further processing.
*   **`return`:** Sends a value back from the function to the calling code. This returned value can be stored in a variable, used in calculations, passed to another function, etc. It does *not* automatically display anything.

### 6.4. Docstrings

*   A string literal (`"""..."""` or `'''...'''`) placed as the first statement in a function's body.
*   Used to explain what the function does, its parameters, and what it returns.
*   Accessible via `function_name.__doc__`.

In [ ]:
def greet(name):
    """Greets the person passed in as a parameter."""
    message = f"Hello, {name}!"
    return message

def calculate_area(length, width):
    """Calculates the area of a rectangle.
    Returns None if length or width is negative.
    """
    if length < 0 or width < 0:
        return None # Indicate invalid input
    return length * width

# Calling functions and using their return values
greeting_msg = greet("Midterm Taker")
print(f"Greeting: {greeting_msg}")

area1 = calculate_area(5, 4)
area2 = calculate_area(7, 3)
invalid_area = calculate_area(-5, 4)

print(f"Area 1 (5x4): {area1}")
print(f"Area 2 (7x3): {area2}")
print(f"Invalid Area (-5x4): {invalid_area}")

print(f"\nDocstring for greet: {greet.__doc__}")

# print vs return example
def print_sum_func(a, b):
    print(a + b) # Prints the sum, returns None implicitly

def return_sum_func(a, b):
    return a + b # Returns the sum

print("\nCalling print_sum(10, 5):")
result_p = print_sum_func(10, 5) 
print(f"Result from print_sum_func: {result_p} (because it only prints)")

result_r = return_sum_func(10, 5)
print(f"\nResult from return_sum_func(10, 5): {result_r}")
print(f"This result can be used: {result_r * 2}")

Greeting: Hello, Midterm Taker!
Area 1 (5x4): 20
Area 2 (7x3): 21
Invalid Area (-5x4): None

Docstring for greet: Greets the person passed in as a parameter.

Calling print_sum(10, 5):
15
Result from print_sum_func: None (because it only prints)

Result from return_sum_func(10, 5): 15
This result can be used: 30


## 7. Quick Review Exercises (Self-Assessment)

*   **Q1:** What is the data type of `x` after `x = "5" + "3"`? What is its value?
    *   *Answer: `str`, value is `"53"` (string concatenation).*

*   **Q2:** Write a slice to get the last 3 characters of a string `s`.
    *   *Answer: `s[-3:]`*

*   **Q3:** What will the following code print?
    ```python
    my_list = [10, 20, 30, 40]
    if 25 in my_list:
        print("Found")
    elif my_list[1] > 15:
        print("Second item is large")
    else:
        print("Not found")
    ```
    *   *Answer: `"Second item is large"` (25 is not in list, `my_list[1]` is 20, which is > 15).*

*   **Q4:** How do you add a key-value pair `("status", "active")` to an existing dictionary `user_profile`?
    *   *Answer: `user_profile["status"] = "active"`*

*   **Q5:** Write a `for` loop that prints numbers from 10 down to 1 (inclusive).
    *   *Answer: `for i in range(10, 0, -1): print(i)`*

*   **Q6:** When would you use a `while` loop instead of a `for` loop?
    *   *Answer: When the number of iterations is not known beforehand and depends on a condition becoming false (e.g., user input, event handling).*

*   **Q7:** What is the main difference between `return` and `print` in a function?
    *   *Answer: `print()` displays output to the console for the user. `return` sends a value back from the function to the calling code, allowing that value to be stored, used in calculations, or passed to other functions.*